### DISCUSSION TITLE
[SOLVED] change the user model

### DISCUSSION THREAD

**Discussion by l-vanel** [codeigniter4/shield#333] [Q&A] [answered]
I want to add some field to the user_identities model and when i read the doc, according the shield concept's, it posiible  by using the user provider. But i don't really understand how to.

**Comment by datamweb** [comment, marked as answer]
Hi @l-vanel,
Thank you for registering your question again in this section.

1. Create a migration file in path `app\Database\Migrations\AddCustomColumnForUser` with the following content:
```php
<?php

namespace App\Database\Migrations;

use CodeIgniter\Database\Migration;

// Add custom column for shield
class AddCustomColumnForUser extends Migration
{
    public function up()
    {
        
        $fields = [
            "gender SET('male', 'female', 'bisexual') NULL DEFAULT NULL COMMENT 'gender user'",
            'first_name' => [
                'type'       => 'text',
                'constraint' => '50',
                'null'       => TRUE
            ],
            'last_name' => [
                'type'       => 'VARCHAR',
                'constraint' => '50',
                'null'       => TRUE
            ],
            'avatar' => [
                'type'       => 'VARCHAR',
                'constraint' => '1000',
                'null'       => TRUE
            ],
            'phone_number' => [
                'type'       => 'VARCHAR',
                'constraint' => '20',
                'null'       => TRUE
            ],
            'company' => [
                'type'       => 'VARCHAR',
                'constraint' => '100',
                'null'       => TRUE
            ],
            'description' => [
                'type'       => 'VARCHAR',
                'constraint' => '1200',
                'null'       => TRUE
            ],
            'code_meli' => [
                'type'       => 'VARCHAR',
                'constraint' => '20',
                'null'       => TRUE
            ],
            'birth_day date default NULL',
            'country' => [
                'type'       => 'VARCHAR',
                'constraint' => '20',
                'null'       => TRUE
            ],        
            'state' => [
                'type'       => 'VARCHAR',
                'constraint' => '20',
                'null'       => TRUE
            ],
            'city' => [
                'type'       => 'VARCHAR',
                'constraint' => '20',
                'null'       => TRUE
            ],
            'full_address' => [
                'type'       => 'VARCHAR',
                'constraint' => '400',
                'null'       => TRUE
            ],
        

        ];

        $this->forge->addColumn('users', $fields);
    
    }

    public function down()
    {
        $fields = [
            'gender',
            'first_name',
            'last_name',
            'company',
            'phone_number',
            'description',
            'code_meli',
            'birth_day',
            'country',
            'state',
            'city',
            'full_address',

        ];
        $this->forge->dropColumn('users', $fields);
    }
}
```
2 . Execute command `php spark migrate` to apply changes to the users table.
3. Checking the addition of columns in table users with command `php spark db:table users`
4. Create file *UserModel.php* via `php spark shield:model` with the following content:
```php 
<?php

declare(strict_types=1);

namespace App\Models;

use CodeIgniter\Shield\Models\UserModel as ShieldUserModel;

class UserModel extends ShieldUserModel
{
    protected function initialize(): void
    {
        parent::initialize();

        $this->allowedFields = [
            ...$this->allowedFields,
          'gender',
          'first_name',
          'last_name',
          'avatar',
          'phone_number',
          'company',
          'code_meli',
          'birth_day ',
          'country',
          'state',
          'city',
          'full_address',

            // 'first_name',
        ];
    }
}
```
6. Edit file app\Config\Auth.php and change as follows:
```php 
    public string $userProvider = \App\Models\UserModel::class;
```
**end of work.**
Now you can use the same methods as in the documentation without any problems.
e.g : 
```php
use CodeIgniter\Shield\Entities\User;

$users = model(\App\Models\UserModel::class);
$user = new User([
    'username' => 'foo-bar',
    'email'    => 'foo.bar@example.com',
    'password' => 'secret plain text password',
    'first_name' => 'Pooya',
    'last_name' => 'Parsa',
]);
$users->save($user);
```
more info [see](https://github.com/codeigniter4/shield/blob/develop/docs/quickstart.md#managing-users)
I believe we should document this in Part [customization](https://github.com/codeigniter4/shield/blob/develop/docs/customization.md) documentation.
Also, thank you very much, after checking this issue, I realized that there are bug in the file `Session.php` , which we will fix in the next few days.

**Reply by solomon-ochepa** [reply]
However, my actual intention was to implement `UUID` in the #users table which I have to alter the table primary key (id) to char(36) and cast the entity key to string, instead of integer. And also disable Auto-increment.

Can you suggest how to get this done with the current design?

**Reply by solomon-ochepa** [reply]
Your suggested solution above is perfect when you just want to add some extra fields/columns to the users' table.

However, you should not be opinionated about the workflow as everyone has their own implementation that may not be as same as yours.

For instance, all our tables are using UUID (including the users' table before integrating the Shield package). Now we are forced to change back to Auto-increment temporarily while working a way out to fall back to UUID again.

Like in the Laravel framework, the migration file should be published to the app migrations dir and allows customization. That will enable easy integration with other packages and workflows.

**Reply by datamweb** [reply]
What you presented in #766 seems off topic.

However, if you think creating a new table can help you, you can use the following options in file **Config/Auth.php**.

https://github.com/codeigniter4/shield/blob/0bf28b24e6f2e7954e67cb686ce214d639812113/src/Config/Auth.php#L72-L80

For **UUID** you can use [this ](https://github.com/michalsn/codeigniter4-uuid).

In general, considering that I have not implemented this issue, I cannot provide an exact solution. However, I believe that by creating a new table, there is a need for changes in other sections.

**Reply by solomon-ochepa** [reply]
`However, I believe that by creating a new table, there is a need for changes in other sections. ...`

You are right, that is the reason why I'm suggesting the migration should be publishable to the app migrations dir and the developer can make all the necessary customization. You can relate this with the "Laravel Roles and Permissions" package.

**Reply by solomon-ochepa** [reply]
`For UUID you can use [this ](https://github.com/michalsn/codeigniter4-uuid).`

This is not needed as we already have our UUID implementation in place.

**Reply by kenjis** [reply]
@solomon-ochepa 
Thank you for your opinion.

> I'm suggesting the migration should be publishable to the app migrations dir and the developer can make all the necessary customization. 

That is a reasonable point, but CI4 does not employ such a mechanism. 
This is probably largely due to practice in the past of not having to use the CLI.

Making such a change is a big change and requires more agreements on the change.

> You can relate this with the "Laravel Roles and Permissions" package.

What do you mean? I don't know what you want to say.

**Reply by kenjis** [reply]
Can you modify the table?
https://codeigniter4.github.io/CodeIgniter4/dbmgmt/forge.html#modifying-a-field-in-a-table

**Reply by CalebArbeau** [reply]
>             "gender SET('male', 'female', 'bisexual') NULL DEFAULT NULL COMMENT 'gender user'",

None of these are genders lol
especially bisexual

**Reply by datamweb** [reply]
I just wanted to show an implementation example to demonstrate the structure. Naturally, you should adjust the values based on your own project needs. The ones I used were just placeholders, not meant to define actual genders or orientations.

Thanks for your attention to detail—hope to see that same sharp eye contribute to improving the official docs too!

**Comment by l-vanel** [comment]
hi , i have one problem , the code work but the data of new customs columns doesn't insert in the  database. in other words ,only the username , email and password are insert in the database when i click on the register

**Reply by datamweb** [reply]
Hey vanel ,
```php
 $users = model('UserModel'); 
```
Change to 
```php
 $users = model('App\Models\UserModel');
```
Did your problem get solved?

**Reply by l-vanel** [reply]
Now my problem is about the insertion of datas

**Reply by datamweb** [reply]
OK, let me test myself by tomorrow. It is now 4:30 am.

**Reply by datamweb** [reply]
> Now my problem is about the insertion of datas

@l-vanel, I know that you are new to this part of GitHub. Therefore, first of all, I will explain the following things to you.
1. You should not use **edit to change the question**. This will make the readers not understand what your question was and why you received such an answer.
2. Try to **ask your questions separately**. For example, the title of this was _change the user model_, but then you made another request.
3. When inserting your code, you should do the following, this will make the reader feel good.(more info [see](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax))
````
```php
// php code 
```
OR 
```html
// html code
```
````

4. There is no need for a personal message, all questions here are reviewed by members. If you do not receive a response, be sure that there was no response time or ability.
_Remember that these explanations are not just for you, these explanations are true for everyone._

**Let's go, to answer your question.**

**Comment by l-vanel** [comment]
Ok, thanks.

**Comment by datamweb** [comment]
Your question is how can I save the values of **gender**, **first_name**, **last_name** and ... in the database in addition to email, username and password on the **registration page.**

**Reply by datamweb** [reply]
**Answer** Do the following steps:

1. Add in the view file(Please use the customization of views)
```html 
<!-- Gender-->
<div class="mb-4">
<select id="gender" class="form-control" name="gender" placeholder="<?= lang('Auth.gender') ?>>
<option value="male">Male</option>
<option value="female">Female</option>
<option value="bisexual">Bisexual</option>
</select>
</div>
<!-- First Name-->
<div class="mb-4">
 <input type="text" class="form-control" name="first_name" inputmode="text" autocomplete="first_name" placeholder="<?= lang('Auth.first_name') ?>" value="<?= old('first_name') ?>" required />
</div>
<!-- Last Name-->
<div class="mb-4">
<input type="text" class="form-control" name="last_name" inputmode="text" autocomplete="last_name" placeholder="<?= lang('Auth.last_name') ?>" value="<?= old('last_name') ?>" required />
</div>
```
2. in file `app/Config/Auth.php` add :
```php
    public array $personalFields = ['gender','first_name', 'last_name'];
```
3. in file `app\Config\Validation.php` add :
```php
       //--------------------------------------------------------------------
    // Rules
    //--------------------------------------------------------------------
    public $registration = [
        'username'         => [
                'required',
                'max_length[30]',
                'min_length[3]',
                'regex_match[/\A[a-zA-Z0-9\.]+\z/]',
                'is_unique[users.username]',
            ],
        'email'            => 'required|max_length[254]|valid_email|is_unique[auth_identities.secret]',
        'gender' => 'in_list[male,female,bisexual]',
        'first_name' => 'required|alpha|min_length[3]|max_length[10]',
        'last_name' => 'required|alpha|min_length[3]|max_length[15]',
        'password'         => 'required|strong_password',
        'password_confirm' => 'required|matches[password]',
    ];
```
Do the same for other fields.

Did I understand your question correctly and was the answer complete?

**Reply by l-vanel** [reply]
yes it's work. thks
